### PR TITLE
core: initialization and Repository

### PR DESCRIPTION
**Pull request by Wondertan**
## Context
The penultimate piece for #30. It introduces initialization flow for the embedded Core node along with repository to access Core' stored data. Basically, it implements some of the important functionality for future `init` command.

## Others
* Implements custom serialization for `core.Config`(rationale is in the code)
* Starts own collection of utilities
* Provides FS, InMem and Mock repositories

## Refs
Depends on #71 and #74
Moves towards done #30

### CODE REVIEW

File: core/config.go [hunk 1]
```diff
@@ -1,13 +1,61 @@
...
+
+func TestConfig(t *testing.T) *Config {
+	cfg := config.ResetTestRoot(t.Name())
+	t.Cleanup(func() {
+		os.RemoveAll(cfg.RootDir)
+	})
+	return cfg
+}
+
+// LoadConfig config from the file under the given 'path'.
```
Comment (renaynay) [inline core/config.go:28]:
```suggestion
// LoadConfig loads the config from the file under the given 'path'.
```

File: core/config.go [hunk 2]
```diff
@@ -1,13 +1,61 @@
...
+		return err
+	}
+	defer f.Close()
+
+	return toml.NewEncoder(f).Encode(cfg)
 }
 
-func TestConfig(tname string) *Config {
-	return config.ResetTestRoot(tname)
+// change default paths to remove unneeded 'config' sub dir
```
Comment (renaynay) [inline core/config.go:54]:
```suggestion
// updateDefaults changes default paths to remove unneeded 'config' sub dir.
```

**Comment by Wondertan** [comment]
Rebased

**Review by liamsi — COMMENTED** [review]
Can you also merge #71 and rebase this on master again?

**Comment by Wondertan** [comment]
@liamsi, need one more approval from @renaynay or @mattdf

**Comment by Wondertan** [comment]
To avoid confusion, I meant an approval not for this PR, but for #71 to rebase this one on master.

**Comment by Wondertan** [comment]
Rebased

**Comment by liamsi** [comment]
> with repository to access Core' stored data

isn't the whole idea that celestia-node only communicates with core (includes accessing data) via the RPC interface (even if local)?

File: core/init.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/celestiaorg/celestia-core/privval"
+	"github.com/celestiaorg/celestia-core/types"
+	tmtime "github.com/celestiaorg/celestia-core/types/time"
+
+	"github.com/celestiaorg/celestia-node/libs/utils"
+)
+
+// defaultValKeyType is default crypto algo for consensus in core
+// TODO(@Wondertan): Should we make this configurable
+const defaultValKeyType = types.ABCIPubKeyTypeSecp256k1
```
Comment (liamsi) [inline core/init.go:18]: No, the configuration of which key type is used during consensus is none of the business of the celestia-node. It should not be configurable here. Remember, we are treating tendermint as a black-box.

File: core/init.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+				PubKey:  pubKey,
+				Power:   10,
+			}},
+		}
+
+		return genDoc.SaveAs(genPath)
+	}
+
+	return nil
+}
```
Comment (liamsi) [inline core/init.go:81]: Why is all this in here? Why does celestia-node need control of this?

**Review by liamsi — COMMENTED** [review]
On first sight it seems this PR breaks somewhat with the design paradigm we have settled on (namely that the the node only knows about very few RPC endpoints and that is the only thing it needs from core). The motivation for that seems to be that the future node-specific data and config are written into the same "repo" as the core node's? 
I understand why this might be desirable but that is really not important for devnet. 

We have to be careful to focus on the DA-specific parts!

File: core/init.go (hunk 4 above)
Comment (Wondertan) [reply core/init.go:81]: > Why is all this in here? Why does celestia-node need control of this?

We embedded the Core by default and `tendermint` sucks in terms of reusability, so we have no choice besides reimplementing this.

File: core/init.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+		params := types.DefaultConsensusParams()
+		params.Validator.PubKeyTypes = []string{defaultValKeyType}
+		genDoc := types.GenesisDoc{
+			ChainID:         fmt.Sprintf("localnet-%v", tmrand.Str(6)),
+			GenesisTime:     tmtime.Now(),
+			ConsensusParams: params,
+			Validators: []types.GenesisValidator{{
+				Address: pubKey.Address(),
+				PubKey:  pubKey,
+				Power:   10,
```
Comment (liamsi) [inline core/init.go:73]: This all seems arbitrary 🤔 Why power == 10? Why do we create a random chainID? 
Is this a result of the decision to embed the core node? I would have liked that all this is not a concern of the node.

**Comment by liamsi** [comment]
> Implements custom serialization for core.Config(rationale is in the code)

Which custom serialzation (toml is not custom)? Where is the rationale?

File: core/init.go (hunk 5 above)
Comment (Wondertan) [reply core/init.go:73]: > This all seems arbitrary :thinking:
It is a stub to make core work somehow. The real genesis should be generated elsewhere(we should provide a tool). I just copied this from TM, as you can't reuse initialization.

**Comment by Wondertan** [comment]
> Remember, we are treating tendermint as a black-box.

Yeeeaaah, on a high level, but the deeper you go, the less it black-boxable 🤷🏻‍♂️

**Comment by Wondertan** [comment]
> The motivation for that seems to be that the future node-specific data and config are written into the same "repo" as the core node's?
I understand why this might be desirable but that is really not important for devnet.

The motivations here are:
* Make embedded Core work
* Simplify deployment of things by having everything related to initialization in one place.
	* This is a crucial part for devnet. Instead of having separate hacky initialization scripts for Core and Node, we control where it actually needs to be controlled. The PR just implements initialization complexity earlier than later.
* Thoroughly understand tendermint initialization flow and its requirement 
* Have higher control over Core

**Comment by Wondertan** [comment]
> Which custom serialzation (toml is not custom)?

Not custom serialization, but custom serialization logic. 

> Where is the rationale?

Next `LoadConfig` func

File: core/init.go (hunk 5 above)
Comment (liamsi) [reply core/init.go:73]: There is a tool: https://github.com/cosmos/cosmos-sdk/blob/4c3aa4dbacf961f90175674e6647f454f45b709a/x/genutil/client/cli/init.go#L71

File: core/config.go [hunk 6]
```diff
@@ -20,3 +27,38 @@ func TestConfig(t *testing.T) *Config {
 	})
 	return cfg
 }
+
+// LoadConfig config from the file under the given 'path'.
+// NOTE: Unfortunately, Core(Tendermint) does not provide us with convenient function to load the Config,
+// so we have to do our own. Furthermore, we have to make own SaveConfig, as the one the Core provides has
+// incompatibility issues with our toml parser version.
```
Comment (liamsi) [inline core/config.go:34]: Please us the same toml version as tendermint master, or, help them update their toml version.

**Comment by liamsi** [comment]
> Yeeeaaah, on a high level, but the deeper you go, the less it black-boxable 🤷🏻‍♂️

Let's still try to stick to that paradigm as much as possible.

File: core/config.go (hunk 6 above)
Comment (liamsi) [reply core/config.go:34]: Seems to already be the case btw: https://github.com/tendermint/tendermint/blob/fb9eaf576afbcfcd83d265d7cc0f7d55f2e59393/go.mod#L6

**Review by renaynay — COMMENTED** [review]
couple nits, I'm okay with this but I hate the idea that we need to do this much work to get core to work embedded within node. It's not that much reward per amount of work that's gone into it so far. It's fine otherwise though. We aim to remove all this regardless after Devnet.

File: core/init.go (hunk 5 above)
Comment (Wondertan) [reply core/init.go:73]: Ok, so we can use it for genesis generation for real networks

**Comment by Wondertan** [comment]
> We aim to remove all this regardless after Devnet.

WDYM?

**Comment by Wondertan** [comment]
According to offsite. We treat Tendermint as a black box for devnet. Later on, we will have to decouple it into more pieces, so this code is likely to stay preserving control over initialization.

**Comment by renaynay** [comment]
I mean, we will not run celestia-node and celestia-core processes together. This is a crutch for devnet. After devnet, all should be communicating via p2p layer (meaning celestia-node does not need to know / care about celestia-core ).

File: core/config.go [hunk 7]
```diff
@@ -4,13 +4,20 @@ import (
...
+// Config is an alias for Core config.
+// It have to be used throughout celestia-node codebase to simplify future config changes and etc.
 type Config = config.Config
 
+// DefaultConfig gives slightly update default config of Core node.
 func DefaultConfig() *Config {
-	return config.DefaultConfig()
+	cfg := config.DefaultConfig()
+	updateDefaults(cfg)
+	return cfg
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply core/config.go:20]: `updateDefaults` does not return anything

File: core/config.go (hunk 6 above)
Comment (Wondertan) [reply core/config.go:34]: Ok, thanks. Resolved?

File: core/init.go (hunk 5 above)
Comment (Wondertan) [reply core/init.go:73]: Resolved?

File: core/init.go (hunk 4 above)
Comment (Wondertan) [reply core/init.go:81]: Resolved?

File: core/init.go (hunk 3 above)
Comment (Wondertan) [reply core/init.go:18]: Ok, removing the todo.

Comment (Wondertan) [reply core/init.go:18]: Resolved?

File: core/config.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (renaynay) [reply core/config.go:20]: Can it return the modified config?

[the remark this answers is not in the document]
Comment (Wondertan) [reply core/config.go:20]: It modifies it as it rcvs a pointer there

**Review by renaynay — APPROVED** [review]
OK Thank you!

File: core/init.go (hunk 5 above)
Comment (liamsi) [reply core/init.go:72]: Yes. We should quickly try to get the node actually running (in tests and then in devnet) with celesia-app to get some actual validation for all this.

**Review by liamsi — APPROVED** [review]
👏🏼 LGMT

**Pull request merged**